controller/api/public/edges.go

```diff
@@ -82,63 +82,99 @@ func (s *grpcServer) getEdges(ctx context.Context, req *pb.EdgesRequest) ([]*pb.
 		return nil, err
 	}
 
-	edge := processEdgeMetrics(inboundResult, outboundResult, resourceType)
+	edge := processEdgeMetrics(inboundResult, outboundResult, resourceType, selectedNamespace)
 	return edge, nil
 }
 
-func processEdgeMetrics(inbound, outbound model.Vector, resourceType string) []*pb.Edge {
+func processEdgeMetrics(inbound, outbound model.Vector, resourceType, selectedNamespace string) []*pb.Edge {
 	edges := []*pb.Edge{}
 	dstIndex := map[model.LabelValue]model.Metric{}
 	srcIndex := map[model.LabelValue][]model.Metric{}
 	resourceReplacementInbound := resourceType
 	resourceReplacementOutbound := "dst_" + resourceType
 
 	for _, sample := range inbound {
-		// skip any inbound results that do not have a client_id, because this means
-		// the communication was one-sided (i.e. a probe or another instance where
-		// the src/dst are not both known) in future the edges command will support
-		// one-sided edges
+		// skip inbound results without a clientID because we cannot construct edge
+		// information
 		if _, ok := sample.Metric[model.LabelName("client_id")]; ok {
 			key := sample.Metric[model.LabelName(resourceReplacementInbound)]
 			dstIndex[key] = sample.Metric
 		}
 	}
 
 	for _, sample := range outbound {
-		// skip any outbound results that do not have a server_id for same reason as
-		// above section
-		if _, ok := sample.Metric[model.LabelName("server_id")]; ok {
-			key := sample.Metric[model.LabelName(resourceReplacementOutbound)]
-			if _, ok := srcIndex[key]; !ok {
-				srcIndex[key] = []model.Metric{}
-			}
-			srcIndex[key] = append(srcIndex[key], sample.Metric)
+		key := sample.Metric[model.LabelName(resourceReplacementOutbound)]
+		if _, ok := srcIndex[key]; !ok {
+			srcIndex[key] = []model.Metric{}
 		}
+		srcIndex[key] = append(srcIndex[key], sample.Metric)
 	}
 
 	for key, sources := range srcIndex {
 		for _, src := range sources {
+			srcNamespace := string(src[model.LabelName("namespace")])
+
 			dst, ok := dstIndex[key]
+
+			// if no destination, build edge entirely from source data
 			if !ok {
-				log.Errorf("missing resource in destination metrics: %s", key)
+				dstNamespace := string(src[model.LabelName("dst_namespace")])
+
+				// skip if selected namespace is given and neither the source nor
+				// destination is in the selected namespace
+				if selectedNamespace != "" && srcNamespace != selectedNamespace &&
```

Comment:
Could this filtering be done in the prometheus query where clause?

Reply:
Thanks for the comment - I was considering narrowing down the Prometheus query, but given that we're still iterating on `linkerd edges`, decided to err on the side of getting more data from Prometheus and filtering after the fact. I want to chat with you and @klingerf next week to see if we can pull any more edge data out of Prometheus, specifically for requests with no identity.